package/kedro_viz/data_access/repositories/modular_pipelines.py

```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
-        self.add_child(
-            modular_pipeline_id,
-            ModularPipelineChild(id=node.id, type=GraphNodeType(node.type)),
-        )
-        return modular_pipeline_id
+    def _add_children_to_parent_pipeline(
+        self,
+        parent_node: ModularPipelineNode,
+        modular_pipeline_id: str,
+        modular_pipeline_inputs_outputs: Set[str],
+    ):
+        """
+        Helper to add modular_pipeline children correctly to
+        parent modular pipelines in case of nesting.
+
+        Here we follow the below rules:
+        - A modular pipeline is a child of it's parent modular pipeline
+        - Any input/output of a modular pipeline is a child of the parent pipeline
+          unless it is an input/output of that parent pipeline or a parameter dataset
+        - Any input/output of a top-level modular pipeline is added as a child
+          to the root modular pipeline
 
-    def has_modular_pipeline(self, modular_pipeline_id: str) -> bool:
-        """Return whether this modular pipeline repository has a given modular pipeline ID.
         Args:
-            modular_pipeline_id: ID of the modular pipeline to check existence in the repository.
-        Returns:
-            Whether the given modular pipeline ID is in the repository.
-        Example:
-            >>> modular_pipelines = ModularPipelinesRepository()
-            >>> modular_pipelines.has_modular_pipeline("__root__")
-            True
-            >>> modular_pipelines.has_modular_pipeline("doesnt exist")
-            False
+            parent_node (ModularPipelineNode): The parent modular pipeline node.
+            modular_pipeline_id (str): The ID of the modular pipeline to add as a child.
+            modular_pipeline_inputs_outputs (Set[str]): A set of inputs/outputs
+                    to/from the modular pipeline
         """
-        return modular_pipeline_id in self.tree
+
+        parent_node.children.add(
+            ModularPipelineChild(
+                id=modular_pipeline_id, type=GraphNodeType.MODULAR_PIPELINE
+            )
+        )
+        for dataset in modular_pipeline_inputs_outputs:
+            if (
+                dataset not in parent_node.inputs
+                and dataset not in parent_node.outputs
+                and dataset not in self.parameters
+            ):
+                parent_node.children.add(
+                    ModularPipelineChild(id=dataset, type=GraphNodeType.DATA)
+                )
+            if dataset not in self.node_mod_pipeline_map:
+                self.node_mod_pipeline_map[dataset] = set()
+            self.node_mod_pipeline_map[dataset].add(modular_pipeline_id)
+
+    def _add_datasets_as_children(
+        self,
+        modular_pipeline: ModularPipelineNode,
+        kedro_node: Node,
+        modular_pipeline_inputs_outputs: Set[str],
+    ):
+        """Helper to add datasets (not parameters) related to task nodes as children.

```

Comment:
I was thinking we could add datasets that have no modular pipeline ID associated with them (and thus belong to the root modular pipeline) in this function instead of in managers.py.